### PR TITLE
fix Sonarqube importer

### PR DESCRIPTION
**Pull request by quirinziessler**
**Description**

This PR is intend to optimize the import of Sonarqube findings via API.

- Added the Issue permalink as reference as suggested in #6834 
- Addded the severity of hotspots to the defectdojo findings instead of hardcoded "Info" #8150 

**Test results**

Ideally you extend the test suite in `tests/` and `dojo/unittests` to cover the changed in this PR.
Alternatively, describe what you have and haven't tested.

Feature is tested and already implemented at our DD instance.

**Documentation**

Please update any documentation when needed in the [documentation folder](https://github.com/DefectDojo/django-DefectDojo/tree/dev/docs))

**Checklist**

This checklist is for your information.

- [x] Make sure to rebase your PR against the very latest `dev`.
- [x] Features/Changes should be submitted against the `dev`.
- [x] Bugfixes should be submitted against the `bugfix` branch.
- [x] Give a meaningful name to your PR, as it may end up being used in the release notes.
- [x] Your code is flake8 compliant.
- [x] Your code is python 3.11 compliant.
- [x] If this is a new feature and not a bug fix, you've included the proper documentation in the docs at https://github.com/DefectDojo/django-DefectDojo/tree/dev/docs as part of this PR.
- [x] Model changes must include the necessary migrations in the dojo/db_migrations folder.
- [x] Add applicable tests to the unit tests.
- [x] Add the proper label to categorize your PR.

### CODE REVIEW

**Comment by Maffooch** [comment]
Good updates. Unit tests will need be updated to accommodate the severity on hotspots. Otherwise looks good to me after all checks as passed

File: dojo/tools/api_sonarqube/importer.py [hunk 1]
```diff
@@ -243,7 +251,16 @@ def import_hotspots(self, test):
                     continue
 
                 issue_type = "SECURITY_HOTSPOT"
-                severity = "Info"
+                if hotspot["vulnerabilityProbability"] == "CRITICAL":
+                    severity = "Medium"
```
Comment (kiblik) [inline dojo/tools/api_sonarqube/importer.py:255]: I suppose, this should not be `"Medium"` :)

File: dojo/tools/api_sonarqube/importer.py [hunk 2]
```diff
@@ -235,6 +242,7 @@ def import_hotspots(self, test):
             logging.info(
                 f'Found {len(hotspots)} hotspots for project {component["key"]}'
             )
+            sonarUrl = client.sonar_api_url.replace("api", "")
```
Comment (kiblik) [inline dojo/tools/api_sonarqube/importer.py:245]: As higher

File: dojo/tools/api_sonarqube/importer.py [hunk 3]
```diff
@@ -260,7 +277,11 @@ def import_hotspots(self, test):
                     )
                 )
                 cwe = self.clean_cwe(rule.get("fixRecommendations", ""))
-                references = self.get_references(
+                try:
+                    sonarqube_permalink = f"[Hotspot permalink]({sonarUrl}security_hotspots?id={hotspot['project']}&hotspots={hotspot['key']}) \n"
+                except:
```
Comment (kiblik) [inline dojo/tools/api_sonarqube/importer.py:282]: As higher

File: dojo/tools/api_sonarqube/importer.py [hunk 4]
```diff
@@ -141,17 +143,22 @@ def import_issues(self, test):
                 rule_id = issue["rule"]
                 rule = client.get_rule(rule_id)
                 severity = self.convert_sonar_severity(issue["severity"])
+                try:
+                    sonarqube_permalink = f"[Issue permalink]({sonarUrl}project/issues?issues={issue['key']}&open={issue['key']}&resolved={issue['status']}&id={issue['project']}) \n"
```
Comment (kiblik) [inline dojo/tools/api_sonarqube/importer.py:147]: What is the purpose of `open` and `resolved` part?
It looks that work also without them.
I do not even know where to click to have `open` in url. How did you get it? I would like to learn something new today :)

**Review by kiblik — CHANGES_REQUESTED** [review]

File: dojo/tools/api_sonarqube/importer.py (hunk 3 above)
Comment (manuel-sommer) [reply dojo/tools/api_sonarqube/importer.py:282]: @kiblik no, these two sonarqube_permalinks are not the same

Comment (kiblik) [reply dojo/tools/api_sonarqube/importer.py:282]: "As higher" was for `except`. URL looks good for me.

File: dojo/tools/api_sonarqube/importer.py (hunk 1 above)
Comment (quirinziessler) [reply dojo/tools/api_sonarqube/importer.py]: fixed it - thanks for checking. Sorry for that...

File: dojo/tools/api_sonarqube/importer.py [hunk 5]
```diff
@@ -124,6 +124,8 @@ def import_issues(self, test):
                 f'Found {len(issues)} issues for component {component["key"]}'
             )
 
+            sonarUrl = client.sonar_api_url.replace("api", "")
```
[the remark this answers is not in the document]
Comment (quirinziessler) [reply dojo/tools/api_sonarqube/importer.py:127]: Got your point. Will change this.

File: dojo/tools/api_sonarqube/importer.py [hunk 6]
```diff
@@ -141,17 +143,22 @@ def import_issues(self, test):
                 rule_id = issue["rule"]
                 rule = client.get_rule(rule_id)
                 severity = self.convert_sonar_severity(issue["severity"])
+                try:
+                    sonarqube_permalink = f"[Issue permalink]({sonarUrl}project/issues?issues={issue['key']}&open={issue['key']}&resolved={issue['status']}&id={issue['project']}) \n"
+                except:
```
[the remark this answers is not in the document]
Comment (quirinziessler) [reply dojo/tools/api_sonarqube/importer.py:148]: Yeah can be changed to this as well to cover best practices. Will do so. Thanks for your input!

File: dojo/tools/api_sonarqube/importer.py (hunk 3 above)
Comment (quirinziessler) [reply dojo/tools/api_sonarqube/importer.py:282]: cp above

File: dojo/tools/api_sonarqube/importer.py (hunk 2 above)
Comment (quirinziessler) [reply dojo/tools/api_sonarqube/importer.py:245]: resolved

File: dojo/tools/api_sonarqube/importer.py (hunk 4 above)
Comment (quirinziessler) [reply dojo/tools/api_sonarqube/importer.py:147]: The required URL structure is really weird. I gave up to understand it to be honest. 

The problem is that if you only enter the ID within the url as when you click around in the UI, sonarqube sometimes throws an error called "the finding can not be displayed as it was not part of the 1000 initial loaded". However if you try to open the finding with the other parameters it works. If you want to share an Issue via the UI, klick on the "get permalink" option and you will receive a similar URL. 

I got the exact solution from [here](https://community.sonarsource.com/t/constructing-issue-link-from-sonarqube-api-response/54719/3)

Comment (quirinziessler) [reply dojo/tools/api_sonarqube/importer.py:147]: I also tried it with less GET parameters but sometimes got the same error so I will leave them in as the data is collected from the API anyway...

Comment (kiblik) [reply dojo/tools/api_sonarqube/importer.py:147]: Ok, in that case, it is strong argument to say: keep it as it is.
Thank for searching it and posting link to SQ forum.

**Review by kiblik — APPROVED** [review]
Looks good for me now. Thank you.

**Comment by damiencarol** [comment]
relaunching a failed unit test to unblock this PR

**Comment by manuel-sommer** [comment]
@damiencarol I don't understand why this one unit test fails to be honest.

**Comment by kiblik** [comment]
Just replace https://github.com/DefectDojo/django-DefectDojo/blob/2a629412e7c9cca33e8459a5c545814ce4c9181a/unittests/tools/test_api_sonarqube_parser.py#L44-L46
by
```python
 tool_conf, _ = Tool_Configuration.objects.get_or_create( 
     name="SQ1_unittests", authentication_type="API", tool_type=tool_type, url='http://dummy.url.foo.bar/api'
 ) 
```

**Review by manuel-sommer — APPROVED** [review]

**Comment by manuel-sommer** [comment]
@Maffooch ready for review.

**Review by Maffooch — APPROVED** [review]

**Comment by manuel-sommer** [comment]
@damiencarol

**Comment by manuel-sommer** [comment]
@blakeaowens and @grendel513 could you approve this PR?

**Review by grendel513 — APPROVED** [review]

**Review by blakeaowens — APPROVED** [review]

**Pull request merged**